### ISSUE TITLE
Bad interaction of many_till! and sep! in nom 4

### ISSUE THREAD

**Issue by Keruspe** [Geal/nom#589]
I have a parser with a section that looks like this:

```
named!(fail<&[u8]>, map!(many_till!(take!(1), ws!(tag!("."))), |(r, _)| r[0]));
```

It triggers this error: "cannot infer type for `E` "

Replacing ws! by any look-alike macro that use sep! triggers the issue. Dropping the ws! or the many_till! prevents hitting the issue.

**Comment by Keruspe** [comment]
```
note: trace_macro
 --> src/main.rs:6:1
  |
6 | named!(fail<&[u8]>, map!(many_till!(take!(1), ws!(tag!("."))), |(r, _)| r[0]));
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: expanding `named! { fail < & [ u8 ] > , map ! (
          many_till ! ( take ! ( 1 ) , ws ! ( tag ! ( "." ) ) ) , | ( r , _ ) | r [ 0 ]
          ) }`
  = note: to `# [ allow ( unused_variables ) ] fn fail < 'a > ( i : & 'a [ u8 ] ) -> $crate
          :: IResult < & 'a [ u8 ] , &[u8] , u32 > {
          map ! (
          i , many_till ! ( take ! ( 1 ) , ws ! ( tag ! ( "." ) ) ) , | ( r , _ ) | r [
          0 ] ) }`
  = note: expanding `map! { i , many_till ! ( take ! ( 1 ) , ws ! ( tag ! ( "." ) ) ) , | ( r , _ ) | r [
          0 ] }`
  = note: to `map ! (
          __impl i , many_till ! ( take ! ( 1 ) , ws ! ( tag ! ( "." ) ) ) ,
          |(r, _)| r[0] ) ;`
  = note: expanding `map! { __impl i , many_till ! ( take ! ( 1 ) , ws ! ( tag ! ( "." ) ) ) ,
          |(r, _)| r[0] }`
  = note: to `{
          pub fn _unify < T , R , F : FnOnce ( T ) -> R > ( f : F , t : T ) -> R {
          f ( t ) } ( many_till ! ( i , take ! ( 1 ) , ws ! ( tag ! ( "." ) ) ) ) . map
          ( | ( i , o ) | { ( i , _unify ( |(r, _)| r[0] , o ) ) } ) }`
  = note: expanding `many_till! { i , take ! ( 1 ) , ws ! ( tag ! ( "." ) ) }`
  = note: to `{
          use :: std :: result :: Result :: * ; use $crate :: { Err , Convert } ; let
          ret ; let mut res = :: std :: vec :: Vec :: new (  ) ; let mut input = i .
          clone (  ) ; loop {
          match ws ! ( input , tag ! ( "." ) ) {
          Ok ( ( i , o ) ) => { ret = Ok ( ( i , ( res , o ) ) ) ; break ; } , _ => {
          match take ! ( input , 1 ) {
          Err ( Err :: Error ( err ) ) => {
          ret = Err (
          Err :: Error ( error_node_position ! ( ErrorKind :: ManyTill , input , err ) )
          ) ; break ; } , Err ( e ) => { ret = Err ( Err :: convert ( e ) ) ; break ; }
          , Ok ( ( i , o ) ) => {
          if i == input {
          ret = Err (
          Err :: Error ( error_position ! ( ErrorKind :: ManyTill , input ) ) ) ; break
          ; } res . push ( o ) ; input = i ; } , } } , } } ret }`

error[E0282]: type annotations needed
 --> src/main.rs:6:1
  |
6 | named!(fail<&[u8]>, map!(many_till!(take!(1), ws!(tag!("."))), |(r, _)| r[0]));
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for `E`
  |
  = note: this error originates in a macro outside of the current crate
```

**Comment by Keruspe** [comment]
By using the expanded code directly I get to this:

```
18 |         match ws!(input, tag!(".")) {
   |               ^^^^^^^^^^^^^^^^^^^^^ cannot infer type for `E`
```

**Comment by Keruspe** [comment]
If I replace ws! by sep! or wrap_sep!, I get the same.
Now I expended wrap_sep in a tmp variable, I got this:

```
#![feature(trace_macros)] 

#[macro_use] extern crate nom;

use std::result::Result::*;
use nom::{Err, Convert, sp};

fn main() {
    let ret;
    let mut res = std::vec::Vec::new(); 
    let i = &b""[..];
    let mut input = i.clone();
    loop{
        let tmp = match sp(input) {
            Err(e) => Err(Err::convert(e)),
            Ok((i1, _)) => {                             
                match tag!(i1, ".") {                                                                     
                    Err(e) => Err(Err::convert(e)),
                    Ok((i2, o)) => {                             
                        match ( sp ) ( i2 ) {                                                                          
                            Err(e) => Err(Err::convert(e)),
                            Ok((i3, _)) => Ok ((i3, o))
                        }
                    }
                }
            }
        };
        match tmp {
            Ok((i, o)) => {
                ret = Ok((i, (res, o)));
                break;
            },
            _ => {
                match take!(input, 1) {
                    Err(Err::Error(err)) => {
                        ret = Err(Err::Error(error_node_position!(ErrorKind::ManyTill,input,err)));
                        break;
                    },
                    Err(e) => {
                        ret = Err(Err::convert(e));
                        break;
                    },
                    Ok((i, o)) => {
                        if i == input {
                            ret = Err(Err::Error(error_position!(ErrorKind::ManyTill,input)));
                            break;
                        }
                        res.push(o);
                        input=i;
                    },
                }
            },
        }
    }
    ret;
}
```

Which outputs this

```
  --> src/main.rs:15:27
   |
14 |         let tmp = match sp(input) {
   |             --- consider giving `tmp` a type
15 |             Err(e) => Err(Err::convert(e)),
   |                           ^^^^^^^^^^^^ cannot infer type for `E`
```

**Comment by Geal** [comment]
it should be fixed by 2d0d8f15e3e8315c

**Comment by Keruspe** [comment]
I confirm it's fixed ! Thanks
